File: exercises/poisson/C++/full_gpu_solution/jacobi.cpp
#include "jacobi.h"

/**
 * Makes a step of the jacobi interaction. Compute the new field, given the old field and the known densitiy field.
*/
void compute_jacobi(field_t * phi_new, field_t * phi_old, field_t * rho, int nFields,grid_t * g)
{
    
    #pragma omp parallel
    {
       #pragma omp single
        {
            
            for(int iField=0 ; iField < nFields ; iField++ )
            {

                auto field_phi_new=phi_new[iField].get_data();
                auto field_phi_old=phi_old[iField].get_data();
                auto field_rho=rho[iField].get_data();
                
                #pragma omp task
                    {
                    /* 
                    DATA MAPPING PER STEP IMPLEMENTATION
                    //#pragma omp target data map(tofrom:g->n,g->dx,g->start,g->end,field_phi_new[0:(g->get_size())],field_phi_old[0:(g->get_size())],field_rho[0:(g->get_size())])
                    */
                    
                   #pragma omp target teams distribute parallel for collapse(2)
                   {
                        for(int i=0; i<g->n[0] ; i++)
                            for( int j=0; j<g->n[1] ; j++)
                            {
                                
                                auto k = g->get_index(i,j);

                                auto nx = g->n[0];
                                auto ny = g->n[1];

                                auto dx = g->dx[0];
                                auto dy = g->dx[1];
                                double aspect2 = (dy/dx)*(dy/dx);

                                field_phi_new[ k ] = 0.5*( (field_phi_old[k - 1] +  field_phi_old[k + 1 ])/(1 + 1./aspect2) + (field_phi_old[k - (ny+2) ] +  field_phi_old[(k + ny+2) ])/(1 + aspect2) - field_rho[k] * dx*dx/( 1 + aspect2   ) );
                            }
                    }
                    }
                }    
            
            }

    }

    

};

File: exercises/poisson/Fortran/full_gpu_solution/jacobi.f90
module jacobi
    use field_module
    implicit none 

    contains 
    subroutine compute_jacobi(phi1, phi2, rho, nFields, grid)
        implicit none 
        integer :: iField, nFields, i, j, Nx, Ny 
        integer :: index, aspect2

        ! Argument declarations
        type(field_t), dimension(:), intent(inout) :: phi1, phi2, rho
        type(grid_t) :: grid
        Nx = grid%n(1)
        Ny = grid%n(2)
        
        !$omp parallel
        !$omp single
        do iField = 1, nFields
            grid = phi1(iField)%grid
            
            !$omp task
            !$omp target teams distribute parallel do private(index,aspect2) collapse(2) 
            !!! map(tofrom:phi2(iField)%data,phi1(iField)%data,rho(iField)%data )
            do i = 2, Ny-1
                do j = 2, Nx-1
                    index = (j + 1) + (i + 1) * (grid%n(2) + 2)
                    aspect2 = (grid%dx(2) / grid%dx(1)) * (grid%dx(2) / grid%dx(1))
                    phi2(iField)%data(index) = 0.5 * ((phi1(iField)%data(index - 1) + phi1(iField)%data(index + 1)) / (1.0_dp + 1.0_dp / aspect2) + (phi1(iField)%data(index - (Ny+2)) + phi1(iField)%data(index + (Ny+2))) / (1.0_dp + aspect2) - rho(iField)%data(index) * grid%dx(1) ** 2 / (1.0_dp + aspect2))
                end do 
            end do
            !$omp end target teams distribute parallel do
            !$omp end task

        end do  
        !$omp end single
        !$omp end parallel

    end subroutine compute_jacobi 

end module jacobi